File: model/src/main/java/pl/agh/harmonytools/model/util/ChordComponentManager.java
package pl.agh.harmonytools.model.util;

import pl.agh.harmonytools.model.ChordComponent;

import java.util.HashMap;

public class ChordComponentManager {

    private static final HashMap<String, ChordComponent> availableChordComponents = new HashMap<>();

    public static ChordComponent chordComponentFromString(String chordComponentString, boolean isDown) {
        if (availableChordComponents.containsKey(chordComponentString + (isDown ? "_down" : ""))) {
            System.out.println("CONTAINS KEY");
            return availableChordComponents.get(chordComponentString + (isDown ? "_down" : ""));
        } else {
            ChordComponent cc = new ChordComponent(chordComponentString, isDown);
            availableChordComponents.put(chordComponentString + (isDown ? "_down" : ""), cc);
            return cc;
        }
    }

    @Deprecated
    public static ChordComponent basicChordComponentFromPitch(Integer pitch, boolean isDown) {
        switch (pitch) {
            case 3:
                return chordComponentFromString("3>", isDown);
            case 4:
                return chordComponentFromString("3", isDown);
            case 5:
                return chordComponentFromString("3<", isDown);
            case 6:
                return chordComponentFromString("5>", isDown);
            case 7:
                return chordComponentFromString("5", isDown);
            case 8:
                return chordComponentFromString("5<", isDown);
        }
        throw new IllegalArgumentException("Unsupported pitch");
    }
}

File: model/src/main/scala/pl/agh/harmonytools/model/util/ChordComponentManager.scala
package pl.agh.harmonytools.model.util

import pl.agh.harmonytools.model.chord.ChordComponent

import scala.collection.immutable.HashMap

object ChordComponentManager {

  private var availableChordComponents: HashMap[(String, Boolean), ChordComponent] = HashMap.empty

  def getRoot: ChordComponent = chordComponentFromString("1")

  def chordComponentFromString(chordComponentString: String, isDown: Boolean = false): ChordComponent = {
    availableChordComponents.find(_._1 == (chordComponentString, isDown)) match {
      case Some((key, value)) => value
      case None =>
        val cc = ChordComponent(chordComponentString, isDown)
        availableChordComponents = availableChordComponents + ((chordComponentString, isDown) -> cc)
        cc
    }
  }

  @deprecated
  def basicChordComponentFromPitch(pitch: Int, isDown: Boolean): ChordComponent = {
    pitch match {
      case 3 => chordComponentFromString("3>", isDown)
      case 4 => chordComponentFromString("3", isDown)
      case 5 => chordComponentFromString("3<", isDown)
      case 6 => chordComponentFromString("5>", isDown)
      case 7 => chordComponentFromString("5", isDown)
      case 8 => chordComponentFromString("5<", isDown)
    }
  }

}
